File: misc/benchmarks/hotcloud_2017/one-col-select.js
bench = require('./bench.js')

/*
var singleColumn = false;
if (singleColumn) {
  var nRows  = 256 * 1024 * 1024 - 1;
  var sizeofUint32 = 4;

  var bytes = nRows * sizeofUint32;

  var a;
  var out;
  var outn;

  var setup = function () {
    a = new Uint32Array(nRows);
    out = new Uint32Array(nRows);
    for (var i = 0; i < nRows; ++i) {
      a[i] = Math.floor(Math.random() * 100);
      out[i] = 0;
    }
    outn = 0;
  };

  var get = function (row, col) {
    return a[row];
  }

  var emit = function (value) {
    out[outn++] = value;
  }

  var reset = function () {
    outn = 0;
  }
} else {
*/
  var fieldSize = 4; // sizeof(Float64)
  var fieldsPerRow = 16;
  var recordSize = fieldsPerRow * fieldSize;
  // Must be strictly less than 1 GB.
  var bytes = 1 * 1024 * 1024 * 1024 - 1;
  var nRows = bytes / recordSize;

  var buf = new ArrayBuffer(bytes);
  var a = new DataView(buf);
  var out = new Uint32Array(nRows);
  var outn;

  var get = function (row, col) {
    return a.getUint32(recordSize * row + fieldSize * col);
  }

  var emit = function (value) {
    out[outn++] = value;
  }

  var set = function (row, col, value) {
    a.setUint32(recordSize * row + fieldSize * col, value);
  }

  var setup = function () {
    for (var i = 0; i < nRows; ++i) {
      for (var j = 0; j < fieldsPerRow; ++j) {
        set(i, j, Math.floor(Math.random() * 100));
      }
      out[i] = 0;
    }
    outn = 0;
  }

  var reset = function () {
    outn = 0;
  }
/*
}
*/

// Schema: A: uint32_t
// Query: select sum(A) from T
function q1() {
  var s = 0;
  for (var i = 0; i < nRows; ++i) {
    s += get(i, 0);
  }
  emit(s);
}

// Schema: A: uint32_t
// Query: select sum(A) from T where A < x
function q2(x) {
  var s = 0;
  for (var i = 0; i < nRows; ++i) {
    var v = get(i, 0);
    if (v < x) {
      s += v;
    }
  }
  emit(s);
}

// Schema: A: uint32_t
// Query: select sum(A) from T where A < x
function q3(x) {
  for (var i = 0; i < nRows; ++i) {
    var v = get(i, 0);
    if (v < x) {
      emit(v + v);
    }
  }
}

// Schema: A: uint32_t
// Query: select A + B + C + D from T where A < x
function q4(x) {
  for (var i = 0; i < nRows; ++i) {
    var a = get(i, 0);
    var b = get(i, 1);
    var c = get(i, 2);
    var d = get(i, 3);
    if (a < x) {
      emit(a + b + c + d);
    }
  }
}

function q5(x) {
  var s = 0;
  for (var i = 0; i < nRows; ++i) {
    var a = get(i, 0);
    var b = get(i, 1);
    var c = get(i, 2);
    var d = get(i, 3);
    if (a < x) {
      s += Math.pow(a, Math.pow(b, Math.pow(c * d)));
    }
  }
  emit(s);
}


//engine.setup();
setup();

bench.start();
q1();
bench.report('OneColSelect::q1', bytes);
reset();

for (var iter = 0; iter < 10; ++iter) {
  for (var selectivity = 0; selectivity < 101; selectivity += 10) {
    bench.start();
    q2(selectivity);
    bench.report('OneColSelect::q2(' + selectivity + ')', bytes);
    reset();

    bench.start();
    q3(selectivity);
    bench.report('OneColSelect::q3(' + selectivity + ')', bytes);
    reset();

    //if (!singleColumn) {
      bench.start();
      q4(selectivity);
      bench.report('OneColSelect::q4(' + selectivity + ')', bytes);
      reset();

      bench.start();
      q5(selectivity);
      bench.report('OneColSelect::q5(' + selectivity + ')', bytes);
      reset();
    //}
  }
}

// select sum(A + B) from T
// select A + B from T
// select A + B from T where A < 10

File: misc/benchmarks/hotcloud_2017/one-col-select.cc
#include "common.h"

#include <random>
#include <atomic>

template <size_t nFields>
struct Relation {
  struct Row {
    std::atomic<uint32_t> fields[nFields];
  };
  static_assert(sizeof(Row) == nFields * sizeof(uint32_t), "GRUU!");

  static constexpr size_t bytes = 1lu * 1024 * 1024 * 1024;
  static constexpr size_t nRows  = bytes / sizeof(Row);

  Row* a;
  std::atomic<uint32_t>* out;
  uint32_t outn;

  Relation()
    : a{new Row[nRows]}
    , out{new std::atomic<uint32_t>[nRows]}
    , outn{}
  {
    std::random_device rd{};
    std::mt19937 gen{rd()};
    std::uniform_int_distribution<> dis(0, 99);

    for (uint32_t i = 0; i < nRows; ++i) {
      for (uint32_t j = 0; j < nFields; ++j)
        a[i].fields[j] = dis(gen);
      out[i] = 0;
    }
  }

  ~Relation() {
    delete[] out;
    delete[] a;
  }

  Relation(const Relation&) = delete;
  Relation& operator=(const Relation&) = delete;

  uint32_t get(uint32_t row, uint32_t col) {
    if (nFields == 1)
      col = 0;
    return a[row].fields[col].load(std::memory_order_relaxed);
  }

  uint32_t set(uint32_t row, uint32_t col, uint32_t value) {
    if (nFields == 1)
      col = 0;
    a[row].fields[col].store(value, std::memory_order_relaxed);
  }

  uint32_t emit(uint32_t value) {
    out[outn++].store(value, std::memory_order_relaxed);
  }

  void reset() {
    outn = 0;
  }
};

Relation<16> r{};

// Schema: A: uint32_t
// Query: select sum(A) from T
void q1() {
  uint32_t s = 0;
  for (uint32_t i = 0; i < r.nRows; ++i) {
    s += r.get(i, 0);
  }
  r.emit(s);
}


// Schema: A: uint32_t
// Query: select sum(A) from T where A < x
void q2(uint32_t x) {
  uint32_t s = 0;
  for (uint32_t i = 0; i < r.nRows; ++i) {
    uint32_t a = r.get(i, 0);
    if (a < x)
      s += a;
  }
  r.emit(s);
}

// Schema: A: uint32_t
// Query: select sum(A) from T where A < x
void q3(uint32_t x) {
  size_t outn = 0;
  for (uint32_t i = 0; i < r.nRows; ++i) {
    uint32_t a = r.get(i, 0);
    if (a < x)
      r.emit(a + a);
  }
}

void q4(uint32_t x) {
  for (uint32_t i = 0; i < r.nRows; ++i) {
    uint32_t a = r.get(i, 0);
    uint32_t b = r.get(i, 1);
    uint32_t c = r.get(i, 2);
    uint32_t d = r.get(i, 3);
    if (a < x) {
      r.emit(a + b + c + d);
    }
  }
}

void q5(uint32_t x) {
  uint32_t s = 0;
  for (uint32_t i = 0; i < r.nRows; ++i) {
    uint32_t a = r.get(i, 0);
    uint32_t b = r.get(i, 1);
    uint32_t c = r.get(i, 2);
    uint32_t d = r.get(i, 3);
    if (a < x) {
      s += pow(a, pow(b, pow(c, d)));
    }
  }
  r.emit(s);
}

int main() {
  auto start = hrc::now();
  q1();
  auto end = hrc::now();
  report("OneColSelect::q1", r.bytes, start, end);
  r.reset();

  for (uint32_t iters = 0; iters < 10; ++iters) {
    for (uint32_t selectivity = 0; selectivity < 101; selectivity += 10) {
      std::string selstr = std::to_string(selectivity);
      start = hrc::now();
      q2(selectivity);
      end = hrc::now();
      report("OneColSelect::q2(" + selstr + ")",
             r.bytes, start, end);
      r.reset();

      start = hrc::now();
      q3(selectivity);
      end = hrc::now();
      report("OneColSelect::q3(" + selstr + ")",
             r.bytes, start, end);
      r.reset();

      start = hrc::now();
      q4(selectivity);
      end = hrc::now();
      report("OneColSelect::q4(" + selstr + ")",
             r.bytes, start, end);
      r.reset();

      start = hrc::now();
      q5(selectivity);
      end = hrc::now();
      report("OneColSelect::q5(" + selstr + ")",
             r.bytes, start, end);
      r.reset();
    }
  }

  return 0;
}

